src/testing/run.ts

```diff
@@ -0,0 +1,282 @@
+    } catch (err) {
+      // Ignore, any errors for these requests are displayed by the CLI server
+    }
+  },
+};
+
+function isPrimitive(value: unknown): boolean {
+  return (
+    typeof value === 'string' ||
+    typeof value === 'number' ||
+    typeof value === 'boolean'
+  );
+}
+
+function makeTestCaseHash<TestCaseType>(
+  testCase: TestCaseType,
+  testCaseHash:
+    | (keyof TestCaseType & string)[]
+    | ((testCase: TestCaseType) => string),
+): string {
+  if (Array.isArray(testCaseHash)) {
+    const concatenated = testCaseHash
+      .map((key) => JSON.stringify(testCase[key]))
+      .join('');
+    return crypto.createHash('md5').update(concatenated).digest('hex');
+  } else {
+    return testCaseHash(testCase);
+  }
+}
+
+async function gatherWithMaxConcurrency(args: {
+  maxConcurrency: number;
+  tasks: (() => Promise<void>)[];
+}): Promise<void> {
+  const promises: {
+    id: string;
+    promise: Promise<{ id: string }>;
+  }[] = [];
+
+  for (const task of args.tasks) {
+    const id = crypto.randomUUID();
+    const promise: Promise<{ id: string }> = (async () => {
+      try {
+        await Promise.resolve(task());
+      } catch (err) {
+        // Ignore, errors are handled in the tasks themselves
+      }
+      return { id };
+    })();
+
+    promises.push({
+      id,
+      promise,
+    });
+
+    if (promises.length >= args.maxConcurrency) {
+      // Remove the first promise that resolves
+      await Promise.race(promises.map((p) => p.promise)).then((winner) => {
+        const toRemove = promises.find((p) => p.id === winner.id);
+        if (toRemove) {
+          promises.splice(promises.indexOf(toRemove), 1);
+        }
+      });
+    }
+  }
+
+  await Promise.allSettled(promises.map((p) => p.promise)); // Ensure all remaining promises are finished
+}
+
+async function sendError(args: {
+  testId: string;
+  testCaseHash: string | null;
+  evaluatorId: string | null;
+  error: unknown;
+}): Promise<void> {
+  if (args.error instanceof Error) {
+    await client.post('/errors', {
+      testExternalId: args.testId,
+      testCaseHash: args.testCaseHash,
+      evaluatorExternalId: args.evaluatorId,
+      error: {
+        name: args.error.name,
+        message: args.error.message,
+        stacktrace: args.error.stack,
+      },
+    });
+  }
+}
+
+async function evaluateOutput<TestCaseType, OutputType>(args: {
+  testId: string;
+  testCase: TestCaseType;
+  testCaseHash: string;
+  output: OutputType;
+  evaluator: BaseTestEvaluator<TestCaseType, OutputType>;
+}): Promise<void> {
+  let evaluation: Evaluation | undefined = undefined;
+
+  try {
+    evaluation = await args.evaluator.evaluateTestCase({
+      testCase: args.testCase,
+      output: args.output,
+    });
+  } catch (err) {
+    await sendError({
+      testId: args.testId,
+      testCaseHash: args.testCaseHash,
+      evaluatorId: args.evaluator.id,
+      error: err,
+    });
+  }
+
+  if (evaluation === undefined) return;
```

Comment:
Instead of having this evaluation method and checking if it's undefined you could just post to evals inside the try block since the evals post is swallowing errors

Reply:
i'm not sure i follow

Reply:
oh i think i see what you're saying

Reply:
```
try {
    const evaluation = await args.evaluator.evaluateTestCase({
      testCase: args.testCase,
      output: args.output,
    });
    await client.post('/evals', {
    testExternalId: args.testId,
    testCaseHash: args.testCaseHash,
    evaluatorExternalId: args.evaluator.id,
    score: evaluation.score,
    threshold: evaluation.threshold,
  });
  } catch (err) {
    await sendError({
      testId: args.testId,
      testCaseHash: args.testCaseHash,
      evaluatorId: args.evaluator.id,
      error: err,
    });
  }
  ```

Reply:
but that would break the catch statement if post stopped swallowing errors so its fine as is

Reply:
oops i already switched it lol

Reply:
haha that works too

Reply:
actually i'm gonna switch it back. the error we want to catch and report is an error in the user-defined evaluate method, not HTTP errors b/w the SDK and the CLI